buildflow/core/app/runtime/actors/pipeline_pattern/pipeline_pool.py

```diff
@@ -124,11 +126,23 @@ async def snapshot(self) -> PipelineProcessorSnapshot:
         replica_snapshots: List[PullProcessPushSnapshot] = []
         # TODO: Dont access self.replicas directly. It should be accessed via a method
         # interface
-        for replica in list(self.replicas.values()):
-            snapshot: PullProcessPushSnapshot = (
-                await replica.ray_actor_handle.snapshot.remote()
-            )
-            replica_snapshots.append(snapshot)
+        dead_replica_idxs = []
+        for i, replica in enumerate(self.replicas):
+            try:
+                snapshot: PullProcessPushSnapshot = (
+                    await replica.ray_actor_handle.snapshot.remote()
+                )
+                replica_snapshots.append(snapshot)
+            except RayActorError:
+                logging.exception("replica actor unexpectedly died. will restart.")
+                # We keep this list reverse sorted so we can iterate and remove
+                dead_replica_idxs.insert(0, i)
```

Comment:
nit: consider using a deque for appendleft

Reply:
done.